File: harness/gitops.py
"""Git worktree isolation: bad runs never dirty the integration branch."""
import os
import subprocess


class GitError(RuntimeError):
    pass


def git(args, cwd):
    r = subprocess.run(["git"] + args, cwd=cwd, capture_output=True, text=True)
    if r.returncode != 0:
        raise GitError(f"git {' '.join(args)}: {r.stderr.strip()}")
    return r.stdout.strip()


def ensure_integration(repo, branch):
    if not git(["branch", "--list", branch], repo):
        git(["branch", branch], repo)


def create_worktree(repo, wt_root, task_id, branch, branch_name=None):
    path = os.path.join(wt_root, task_id)
    bname = branch_name or f"harness/task/{task_id}"
    # -B not -b: blocked tasks keep their branch for audit, and retries
    # (alt-model, rerun) must reset it rather than crash on "already exists"
    git(["worktree", "add", "-B", bname, path, branch], repo)
    return path


def commit_all(wt, msg):
    git(["add", "-A"], wt)
    r = subprocess.run(["git", "commit", "-m", msg], cwd=wt,
                       capture_output=True, text=True)
    # nonzero with "nothing to commit" is fine; anything else is real
    if r.returncode != 0 and "nothing to commit" not in r.stdout + r.stderr:
        raise GitError(f"commit: {r.stderr.strip() or r.stdout.strip()}")


def changed_paths(wt, base_branch):
    out = git(["diff", "--name-only", base_branch], wt)
    return [l for l in out.splitlines() if l.strip()]


def diff_text(wt, base_branch):
    return git(["diff", base_branch], wt)


def integration_worktree(repo, wt_root, branch):
    path = os.path.join(wt_root, "_integration")
    if not os.path.exists(path):
        git(["worktree", "add", path, branch], repo)
    return path


def merge_task(int_wt, task_id, branch_name=None):
    bname = branch_name or f"harness/task/{task_id}"
    git(["merge", "--no-ff", "-m", f"harness: ship {task_id}",
         bname], int_wt)


def remove_worktree(repo, path, task_id, keep_branch, branch_name=None):
    git(["worktree", "remove", "--force", path], repo)
    if not keep_branch:
        bname = branch_name or f"harness/task/{task_id}"
        git(["branch", "-D", bname], repo)


def stale_worktrees(wt_root):
    if not os.path.isdir(wt_root):
        return []
    return [os.path.join(wt_root, d) for d in os.listdir(wt_root)
            if d != "_integration"]

File: src/server/harness/gitops.ts
/** Full port of backend/harness/gitops.py */
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";

export class GitError extends Error {}

export function git(args: string[], cwd: string): string {
  const r = spawnSync("git", args, { cwd, encoding: "utf-8" });
  if (r.status !== 0) {
    throw new GitError(`git ${args.join(" ")}: ${(r.stderr || r.stdout || "").trim()}`);
  }
  return (r.stdout || "").trim();
}

export function ensureIntegration(repo: string, branch: string): void {
  if (!git(["branch", "--list", branch], repo)) git(["branch", branch], repo);
}

export function createWorktree(
  repo: string,
  wtRoot: string,
  taskId: string,
  branch: string,
  branchName?: string,
): string {
  const wtPath = path.join(wtRoot, taskId);
  const bname = branchName ?? `harness/task/${taskId}`;
  git(["worktree", "add", "-B", bname, wtPath, branch], repo);
  return wtPath;
}

export function commitAll(wt: string, msg: string): void {
  git(["add", "-A"], wt);
  const r = spawnSync("git", ["commit", "-m", msg], { cwd: wt, encoding: "utf-8" });
  const out = (r.stdout || "") + (r.stderr || "");
  if (r.status !== 0 && !out.includes("nothing to commit")) {
    throw new GitError(`commit: ${out.trim()}`);
  }
}

export function changedPaths(wt: string, baseBranch: string): string[] {
  const out = git(["diff", "--name-only", baseBranch], wt);
  return out.split(/\r?\n/).filter((l) => l.trim());
}

export function diffText(wt: string, baseBranch: string): string {
  return git(["diff", baseBranch], wt);
}

export function integrationWorktree(repo: string, wtRoot: string, branch: string): string {
  const wtPath = path.join(wtRoot, "_integration");
  if (!fs.existsSync(wtPath)) git(["worktree", "add", wtPath, branch], repo);
  return wtPath;
}

export function mergeTask(intWt: string, taskId: string, branchName?: string): void {
  const bname = branchName ?? `harness/task/${taskId}`;
  git(["merge", "--no-ff", "-m", `harness: ship ${taskId}`, bname], intWt);
}

export function removeWorktree(
  repo: string,
  wtPath: string,
  taskId: string,
  keepBranch: boolean,
  branchName?: string,
): void {
  git(["worktree", "remove", "--force", wtPath], repo);
  if (!keepBranch) {
    const bname = branchName ?? `harness/task/${taskId}`;
    git(["branch", "-D", bname], repo);
  }
}

export function staleWorktrees(wtRoot: string): string[] {
  if (!fs.existsSync(wtRoot)) return [];
  return fs
    .readdirSync(wtRoot)
    .filter((d) => d !== "_integration")
    .map((d) => path.join(wtRoot, d));
}

export function recover(cfg: Record<string, unknown>, onlyTaskIds?: Set<string>): void {
  const wtRoot = String(cfg.worktree_root);
  const repo = String(cfg.repo_root);
  for (const wtPath of staleWorktrees(wtRoot)) {
    const name = path.basename(wtPath);
    if (onlyTaskIds && !onlyTaskIds.has(name)) continue;
    try {
      git(["worktree", "remove", "--force", wtPath], repo);
    } catch {
      fs.rmSync(wtPath, { recursive: true, force: true });
    }
  }
  try {
    git(["worktree", "prune"], repo);
  } catch {
    /* ignore */
  }
}
